web/api/rest-api-design/Error-Response-Strategy.md

```diff
@@ -1,44 +1,97 @@
 
-### Strategy on returning consistent errors across the API
+### Use a consistent error format across the API
+
+When the client handles an error, they should know the format of the error beforehand. To make this simple, use the same error format across all errors, at least at the top level.
+
+This way the error handling code on the frontend can be somewhat generic, instead of having to adapt to different response types based on the API.
+
+**Example**
+A bad example would be to return errors like these in the same system. Notice one has parameters `Message` and `MissingParameter` and the other `Code` and `Message`:
+
+```json
+  "Message": "Required parameter AppId is missing.",
+  "MissingParameter": "SearchRequest.AppId",
+```
+
+```json
+  "Code": "125764",
+  "Message": "Transaction could not be completed"
+```
+
+The client cannot tell whether `code` will always be returned or not, they would have to match the string. The string could change in another similar error.
+
+A good example would be to return an error like this, where the `Code`, `Title` and `Message` are available in ALL errors across the API:
+
+```json
+  "Code": "VALIDATION_ERROR",
+  "Title": "Validation Error",
+  "Message": "Required parameter is missing from body"
+  "Body": {
+    "Parameter": "appId"
+  }
+```
+
+There's an argument in favor of returning a list of error objects, rather than a single error object, so that the frontend doesn’t have to check every time what the shape of the error response is. This is especially true if you may return multiple errors anywhere in the API.
+
+### Use the correct HTTP status codes
+
+HTTP defines a lot of [meaningful status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) that can be returned from an API. Use these as your HTTP response type.
+
+For e.g. The types of responses are grouped into the following top level categories.
 
-* You should always return consistent errors. This way your error
-  handling code on the front end can be somewhat generic, instead of
-  having to adapt to different response types based on the API.
+* `2xx` **Success** - This category indicates the action requested by the client was received, understood, and accepted.
+* `4xx` **Client Error** - This category indicate that the error seems to have been caused by the client.
+* `5xx` **Server Error** - This category indicates the server failed to fulfill a request.
 
-* This is an argument in favor of returning a list of error objects,
-  rather than a single error object, so that the front end doesn’t
-  have to check every time what the shape of the error response is
+**Importance of the Client vs. Server Error Distinction** - `500` server errors signal that something went wrong on the backend *that shouldn't have*. Most known errors should be a `4xx`. Most `500` errors are bugs. Keep this in mind as you select which type of error response your situation calls for.
```

Comment:
Yes, that's why I said "something went wrong on the backend"? Perhaps that's not clear?

The distinction I'm trying to make here is that's not enough that it's "a server side error". What I'm seeing is if devs have any sort of error handling a request, they return a 500 because they detected the error in the server code. If the client sent a bad value and it triggered an error in the backend, it's a 500. 

But what we want is the source of the error, the thing that went wrong, should be the server, not the contents of the request. I'll try an update to make that clearer if it isn't.

Reply:
Oh... I guess I was commenting on using the singular `500`, rather than `5XX` to denote all 500-series codes.

Reply:
Ahh I see.